xml-conduit/src/Text/XML/Stream/Parse.hs

```diff
@@ -227,13 +227,48 @@ tokenToEvent _ es n (TokenEndElement name) =
         case n of
             []   -> (NSLevel Nothing Map.empty, [])
             x:xs -> (x, xs)
-tokenToEvent _ es n (TokenContent (ContentEntity e))
-    | Just t <- lookup e es = (es, n, [EventContent $ ContentText t])
+tokenToEvent ps es n (TokenContent (ContentEntity e))
+    = (es, n, map EventContent (resolveEntities ps es [ContentEntity e]))
 tokenToEvent _ es n (TokenContent c) = (es, n, [EventContent c])
 tokenToEvent _ es n (TokenComment c) = (es, n, [EventComment c])
 tokenToEvent _ es n (TokenDoctype t eid es') = (es ++ es', n, [EventBeginDoctype t eid, EventEndDoctype])
 tokenToEvent _ es n (TokenCDATA t) = (es, n, [EventCDATA t])
 
+resolveEntities :: ParseSettings
+                -> [(Text, Text)]  -- entity table
+                -> [Content]
+                -> [Content]
+resolveEntities ps entities = foldr go []
+ where
+  go c@(ContentEntity e) cs
+    = case expandEntity entities e of
+        Just xs -> foldr go cs xs
+        Nothing ->  c : cs
+  go c cs = c:cs
+  expandEntity es e
+    | Just t <- lookup e es =
+      case AT.parseOnly (manyTill
+                          (parseContent ps False False :: Parser Content)
+                          AT.endOfInput) t of
+        Left _      -> Nothing
+        Right xs    ->
+          -- we delete e from the entity map in resolving its contents,
+          -- to avoid infinite loops in recursive expansion:
+          case foldr (goent (filter (\(x,_) -> x /= e) es)) (Just ([], 0)) xs of
+            Just (cs, _) -> Just cs
+            Nothing -> Nothing
+    | otherwise     = Nothing
+  goent _ _ Nothing = Nothing
+  goent es c@(ContentEntity e) (Just (cs, size))
+    = case expandEntity es e of
+        Just xs -> foldr (goent es) (Just (cs, size)) xs
+        Nothing -> Nothing
```

Comment:
Similarly, I believe this can be simplified into:
```haskell
foldr (goent es) (Just (cs, size)) <$> expandEntity es e
```

Reply:
Actually it needs to be
```
expandEntity es e >>= foldr etc.
```
I'll write it that way if you prefer that style.

Reply:
Right, my bad.